indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/symmetric/SymmetricEncryptionAlgorithm.kt

```diff
@@ -0,0 +1,527 @@
+    }
+}
+
+/**
+ * Defines whether a cipher is authenticated or not
+ */
+sealed interface AuthCapability<K : KeyType> {
+    /**
+     * accessor to get the key type due to type erasure
+     */
+    val keyType: K
+
+    /**
+     * Indicates an authenticated cipher
+     */
+    sealed class Authenticated<K : KeyType>(val tagLength: BitLength, override val keyType: K) : AuthCapability<K> {
+
+        /**
+         * An authenticated cipher construction that is inherently authenticated and thus permits no dedicated MAC key
+         */
+        class Integrated(tagLen: BitLength) : Authenticated<KeyType.Integrated>(tagLen, KeyType.Integrated)
+
+        /**
+         * An authenticated cipher construction based on an unauthenticated cipher with a dedicated MAC function, requiring a dedicated MAC key.
+         * _Encrypt-then-MAC_
+         */
+        class WithDedicatedMac<M : MAC, I : NonceTrait>(
+            /**
+             * The inner unauthenticated cipher
+             */
+
+            val innerCipher: SymmetricEncryptionAlgorithm<Unauthenticated, I, KeyType.Integrated>,
+            /**
+             * The mac function used to provide authenticated encryption
+             */
+            val mac: M,
+
+            /**
+             * The preferred length pf the MAC key. Can be overridden during key generation and is unconstrained.
+             */
+            val preferredMacKeyLength: BitLength,
+
+            tagLen: BitLength,
+
+            /**
+             * Specifies how the inputs to the MAC are to be encoded/processed
+             */
+            val dedicatedMacInputCalculation: DedicatedMacInputCalculation,
+
+            /**
+             * Specifies how the inputs to the MAC are to be encoded/processed
+             */
+            val dedicatedMacAuthTagTransform: DedicatedMacAuthTagTransformation
+        ) : Authenticated<KeyType.WithDedicatedMacKey>(tagLen, KeyType.WithDedicatedMacKey) {
+
+            override fun equals(other: Any?): Boolean {
+                if (this === other) return true
+                if (other !is WithDedicatedMac<*, *>) return false
+                if (!super.equals(other)) return false
+
+                if (innerCipher != other.innerCipher) return false
+                if (mac != other.mac) return false
+                if (preferredMacKeyLength != other.preferredMacKeyLength) return false
+                if (dedicatedMacInputCalculation != other.dedicatedMacInputCalculation) return false
+
+                return true
+            }
+
+            override fun hashCode(): Int {
+                var result = super.hashCode()
+                result = 31 * result + innerCipher.hashCode()
+                result = 31 * result + mac.hashCode()
+                result = 31 * result + preferredMacKeyLength.hashCode()
+                result = 31 * result + dedicatedMacInputCalculation.hashCode()
+                return result
+            }
```

Comment:
are these overrides not simply `data class`?

Reply:
since when do you trust those?

Reply:
since there are no bytearrays anywhere in sight

Reply:
better safe than sorry?!

Reply:
better default than something you need to manually maintain